pennylane/transforms/transpile.py

```diff
@@ -0,0 +1,156 @@
+
+    **Example**
+
+    Consider the following example circuit
+
+    .. code-block:: python
+
+        def circuit():
+            qml.CNOT(wires=[0, 1])
+            qml.CNOT(wires=[2, 3])
+            qml.CNOT(wires=[1, 3])
+            qml.CNOT(wires=[1, 2])
+            qml.CNOT(wires=[2, 3])
+            qml.CNOT(wires=[0, 3])
+            return qml.probs(wires=[0, 1, 2, 3])
+
+    which, before transpiling it looks like this:
+
+        0: ──╭C──────────────╭C──╭┤ Probs
+        1: ──╰X──╭C──╭C──────│───├┤ Probs
+        2: ──╭C──│───╰X──╭C──│───├┤ Probs
+        3: ──╰X──╰X──────╰X──╰X──╰┤ Probs
+
+    Suppose we have a device which has connectivity constraints according to the graph:
+
+        0 --- 1
+        |     |
+        2 --- 3
+
+    We encode this in a coupling map as a list of the edges which are present in the graph, and then pass this, together
+    with the circuit, to the transpile function to get a circuit which can be executed for the specified coupling map
+
+    >>> dev = qml.device('default.qubit', wires=[0, 1, 2, 3])
+    >>> transpiled_circuit = qml.transforms.transpile(coupling_map=[(0, 1), (1, 3), (3, 2), (2, 0)])(circuit)
+    >>> transpiled_qnode = qml.QNode(circuit, dev)
+    >>> print(qml.draw(transpiled_qnode))
+
+     0: ──╭C─────────────────────╭C──╭┤ Probs
+     1: ──╰X──╭C─────────╭C──────│───├┤ Probs
+     2: ──╭C──│───╭SWAP──│───╭X──╰X──├┤ Probs
+     3: ──╰X──╰X──╰SWAP──╰X──╰C──────╰┤ Probs
+
+    A swap gate has been applied to wires 2 and 3, and the remaining gates have been adapted accordingly
+
+    """
+    # init connectivity graph
+    coupling_graph = nx.Graph(coupling_map) if not isinstance(coupling_map, nx.Graph) else coupling_map
+
+    # make sure every wire is present in coupling map
+    if any([wire not in coupling_graph.nodes for wire in tape.wires]):
+        raise ValueError(
+            f'not all wires present in coupling map! wires: {tape.wires}, coupling map: {coupling_graph.nodes}')
+
+    def _adjust_op_indices(_op, _map_wires):
+        """ helper function which adjusts wires in Operation according to the map _map_wires"""
+        _new_wires = Wires([_map_wires[w] for w in _op.wires])
+        _params = _op.parameters
+        if len(_params) == 0:
+            return type(_op)(wires=_new_wires)
+        return type(_op)(*_params, wires=_new_wires)
+
+    def _adjust_mmt_indices(_m, _map_wires):
+        """ helper function which adjusts wires in MeasurementProcess according to the map _map_wires"""
+        _new_wires = Wires([_map_wires[w] for w in _m.wires])
+
+        if _m.obs is None:
+            return type(_m)(return_type=_m.return_type, eigvals=_m.eigvals, wires=_new_wires)
+
+        # change wires of observable
+        _params = _m.obs.parameters
+        if len(_params) == 0:
+            _new_obs = type(_m.obs)(wires=_new_wires, id=_m.obs.id)
+        else:
+            _new_obs = type(_m.obs)(_params, wires=_new_wires, id=_m.obs.id)
```

Comment:
Similarly, this edge case does not appear to be tested

Reply:
I wonder if this works in the case of `return qml.expval(H)` where `H = qml.Hamiltonian` 🤔

Reply:
it actually doesn't, I missed this point in the implementation. Same for tensor products, e.g. `return [qml.expval(qml.PauliZ(0) @ qml.PauliX(1))]` . There needs to be put more work into this I think, so that the function `_adjust_mmt_indices` can handle these cases as well.

Reply:
Yes, perhaps the best approach then is to do the following:

1. We can open an issue detailing the need to be able to easily map the wires of tensor product/Hamiltonian objects

2. In this PR, we add a `.. warning::` box near the top of the docstring, informing readers that this function does not yet work for tensors/Hamiltonians.

3. Within the code, we raise an error (for now) in those two cases.

Reply:
ok, great!

Reply:
I will include this in the next commit

Reply:
I think the "else" statement here which does not get tested might be redundant if we catch hamiltonians before; i.e. if we have an observable with parameters, then this gets cast to a Hamiltonian, is that correct? (and then this part cannot get executed and also not tested)

Reply:
> I think the "else" statement here which does not get tested might be redundant if we catch hamiltonians before

Yes, I think so too :)